### DISCUSSION TITLE
High Memory consumption on query node

### DISCUSSION THREAD

**Discussion by Satyadev592** [milvus-io/milvus#32570] [Q&A and General discussion]
Milvus Version: 2.2.9
Number of entries: 115 Million
Dimesnion: 256 
Index: HNSW
Query Nodes: 19 nodes with 64GB RAM and 8 cores CPU. 

We have been using milvus for over a year in production with minimal issues. We have always seen query node memory increasing over a period of few weeks. We restarted the query nodes to resolve this and were doing this periodically. Recently we added a new collection to the same cluster (although a much smaller collection) ~ 4-5 Million entries (~4% size of the other collection). We are noticing frequent memory spikes as a result of this.

All graphs are for last 1 day: 

Query Node Memory Utilisation: 
<img width="1469" alt="image" src="https://github.com/milvus-io/milvus/assets/12812275/02854227-f0fa-4165-a980-95caed6ad69b">

Segment Number by Segment Stage Graph: 
<img width="1458" alt="image" src="https://github.com/milvus-io/milvus/assets/12812275/f34c758a-0083-4113-a910-ae60e3c581b0">

Kindly help us debug and resolve this.

**Comment by xiaofan-luan** [comment]
is there any insertions or deletion on that new collection?

**Reply by Satyadev592** [reply]
Yes, there are insertions and deletions happening on the cluster continuously. Here are the graphs for this: 

Insertions (For last 1 week): 
<img width="1473" alt="image" src="https://github.com/milvus-io/milvus/assets/12812275/e436e1be-1df3-4396-a753-d778d8c0dcfb">

For deletions (For last 1 week): 
<img width="1482" alt="image" src="https://github.com/milvus-io/milvus/assets/12812275/47bcfde2-eed4-4cc3-ab32-500b4f938abe">

**Reply by yhmo** [reply]
Seems there are large amount of continual insert/delete actions every day. The segment number quickly increases from 900 to 1200 and then decreases to 700 in a short time, with an interval around 17 hours.

I think this phenomenon makes sense according to the internal logic of Milvus to process continually insert/delete actions.

The number of delete actions is almost equal to the number of insert actions. I believe the insert/delete actions are for "update" actions.

Milvus stores the newly inserted entities into new segments, and marks the "deleted" entities in the old segments. The "deleted" entities are not erased immediately. The search requests ignore these "deleted" entities so that they are invisible to users, but they are still kept in memory.

Only when the number of deleted entities exceeds a threshold(default 10%) in a segment, Milvus GC process will trigger "self-compaction" actions to this kind of segments. A "self-compaction" to the following things:
1. copy the non-deleted entities into a new segment and persist.
2. mark the original segment to be "deleted", the original segment will wait GC process to delete.
3. build index for the new segment.
4. once the index is ready, load the index of the new segment into query node's memory, and erase the original segment's index from memory.

I guess the "update" actions are averagely distributed to the 800 segments, and the newly inserted entities generate 100 new segments. The segment number increases to 900. At 17:00PM of Tue Apr 23, there are 300 segments reach 10% deletion almost at the same hour, the GC process begins to do "self-compaction" for the segments. 300 new segments are generated. And the 300 old segments are marked as "deleted". The number of segments increases to 1200. After a few minutes, the indexes are ready. Milvus loads the new indexes into memory to replace the original indexes. Since the deleted entities are erased, the memory usage decreases from 60GB to 30+GB.

With continually insert/delete, the memory usage increases to 60GB again at 10:00AM of Wed Apr 24. "self-compaction" triggers again.

**Reply by yhmo** [reply]
A strange point is: 115M 256dim, the original data size is 115GB. With replica_num=2, ideally each query node should take 12GB to load the collection. But the memory usage is 30+GB.

**Reply by Satyadev592** [reply]
Would compacting more often might result in spikes going away? I remember no parameters being there to control compaction automatically and the only way here would be to schedule a daily compaction. Given our high volume of deletes - maybe that makes sense?

**Reply by yhmo** [reply]
In fact, compaction process checks the segments with an interval(1 minute), pretty often. The graph of segment number shows the compaction process almost did nothing during 18:00PM of Apr 23 to 6:00Am of Apr 24, which indicates no segments meet the conditions to do compaction.

**Reply by yhmo** [reply]
Seems the "milvus_querynode_segment_num(*) by {segment_state}" counts all segments including dropped/growing/sealed. You can specify the segment_state respectively to observe the number of different states of segments.

**Reply by Satyadev592** [reply]
The above graph contains two lines - one for sealed and the other for growing. Uploading image of the same graph with both lines clearly highlighted. I don't however see any info on deleted. 

![image](https://github.com/milvus-io/milvus/assets/12812275/7e3f708e-5d3a-4175-8d8e-8f975ca5fcdf)

**Reply by yhmo** [reply]
How many partitions in the collection? 400?

**Reply by Satyadev592** [reply]
Also, do note that the number of segments decreasing from 1200 to 700 is because we issued a restart of all query nodes. I am not sure if your analysis that it's because of compaction holds true there. The number of partitions is 504.

**Reply by yhmo** [reply]
Could you show me the "flushed" and "dropped" segments curve?

**Comment by yhmo** [comment]
115M 256dim, the original data size is 115GB.
HNSW index size is a bit larger than the original data size.
Ideally, the memory consumption of each query node should be 115/19 = 6GB.
From the screenshot, seems each query node's memory usage is more than 30GB.

Did you set replica_num when you call the load() interface?
Are there any scalar fields with long-length varchar in the collection?

**Reply by Satyadev592** [reply]
There are no scalar fields, just the id and embedding. Yes, replica number is set to 2. 

<img width="1739" alt="image" src="https://github.com/milvus-io/milvus/assets/12812275/8806d2cc-b350-4c1a-858b-4e8e99a47da7">
